File: padlock-purge/src/main/java/com/pyamsoft/padlock/purge/PurgePresenter.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.purge;

import android.support.annotation.NonNull;
import com.pyamsoft.pydroid.bus.EventBus;
import com.pyamsoft.pydroid.presenter.SchedulerPresenter;
import io.reactivex.Scheduler;
import javax.inject.Inject;
import javax.inject.Named;
import timber.log.Timber;

class PurgePresenter extends SchedulerPresenter {

  @NonNull private final PurgeInteractor interactor;

  @Inject PurgePresenter(@NonNull PurgeInteractor interactor, @Named("obs") Scheduler obsScheduler,
      @Named("sub") Scheduler subScheduler) {
    super(obsScheduler, subScheduler);
    this.interactor = interactor;
  }

  /**
   * public
   */
  void registerOnBus(@NonNull PurgeCallback callback) {
    //disposeOnStop(EventBus.get()
    //    .listen(PurgeEvent.class)
    //    .subscribeOn(getBackgroundScheduler())
    //    .observeOn(getForegroundScheduler())
    //    .subscribe(purgeEvent -> callback.purge(purgeEvent.packageName()),
    //        throwable -> Timber.e(throwable, "onError purge single")));
    //
    //disposeOnStop(EventBus.get()
    //    .listen(PurgeAllEvent.class)
    //    .subscribeOn(getBackgroundScheduler())
    //    .observeOn(getForegroundScheduler())
    //    .subscribe(purgeAllEvent -> callback.purgeAll(),
    //        throwable -> Timber.e(throwable, "onError purge all")));
  }

  /**
   * public
   */
  void retrieveStaleApplications(@NonNull RetrievalCallback callback, boolean forceRefresh) {
    disposeOnStop(interactor.populateList(forceRefresh)
        .subscribeOn(getBackgroundScheduler())
        .observeOn(getForegroundScheduler())
        .doAfterTerminate(callback::onRetrievalComplete)
        .subscribe(callback::onStaleApplicationRetrieved,
            throwable -> Timber.e(throwable, "onError retrieveStaleApplications")));
  }

  /**
   * public
   */
  void deleteStale(@NonNull String packageName, @NonNull DeleteCallback callback) {
    disposeOnStop(interactor.deleteEntry(packageName)
        .subscribeOn(getBackgroundScheduler())
        .observeOn(getForegroundScheduler())
        .subscribe(() -> callback.onDeleted(packageName),
            throwable -> Timber.e(throwable, "onError deleteStale")));
  }

  interface PurgeCallback {

    void purge(@NonNull String packageName);

    void purgeAll();
  }

  interface RetrievalCallback {

    void onStaleApplicationRetrieved(@NonNull String name);

    void onRetrievalComplete();
  }

  interface DeleteCallback {

    void onDeleted(@NonNull String packageName);
  }
}

File: padlock-purge/src/main/java/com/pyamsoft/padlock/purge/PurgePresenter.kt
/*
 * Copyright (C) 2018 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.purge

import androidx.lifecycle.Lifecycle.Event.ON_STOP
import com.pyamsoft.padlock.api.PurgeInteractor
import com.pyamsoft.padlock.model.purge.PurgeAllEvent
import com.pyamsoft.padlock.model.purge.PurgeEvent
import com.pyamsoft.pydroid.core.bus.EventBus
import com.pyamsoft.pydroid.core.presenter.Presenter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

@JvmSuppressWildcards
class PurgePresenter @Inject internal constructor(
  private val interactor: PurgeInteractor,
  private val purgeBus: EventBus<PurgeEvent>,
  private val purgeAllBus: EventBus<PurgeAllEvent>
) : Presenter<PurgePresenter.View>() {

  override fun onCreate() {
    super.onCreate()
    registerOnBus()
  }

  override fun onStart() {
    super.onStart()
    retrieveStaleApplications(false)
  }

  private fun registerOnBus() {
    dispose {
      purgeBus.listen()
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe({ view?.onPurge(it.packageName) }, {
            Timber.e(it, "onError purge single")
          })
    }

    dispose {
      purgeAllBus.listen()
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe({ view?.onPurgeAll() }, {
            Timber.e(it, "onError purge all")
          })
    }
  }

  fun retrieveStaleApplications(force: Boolean) {
    dispose(ON_STOP) {
      interactor.fetchStalePackageNames(force)
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .doAfterTerminate { view?.onRetrieveComplete() }
          .doOnSubscribe { view?.onRetrieveBegin() }
          .subscribe({ view?.onRetrievedList(it) }, {
            Timber.e(it, "onError retrieveStaleApplications")
            view?.onRetrieveError(it)
          })
    }
  }

  fun deleteStale(packageName: String) {
    dispose {
      interactor.deleteEntry(packageName)
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe({ view?.onDeleted() }, {
            Timber.e(it, "onError deleteStale")
          })
    }
  }

  fun deleteStale(packageNames: List<String>) {
    dispose {
      interactor.deleteEntries(packageNames)
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe({ view?.onDeleted() }, {
            Timber.e(it, "onError deleteStale all")
          })
    }
  }

  interface View : BusCallback, DeleteCallback, RetrieveCallback

  interface RetrieveCallback {

    fun onRetrieveBegin()

    fun onRetrieveComplete()

    fun onRetrievedList(result: List<String>)

    fun onRetrieveError(throwable: Throwable)
  }

  interface DeleteCallback {

    fun onDeleted()
  }

  interface BusCallback {

    fun onPurge(packageName: String)

    fun onPurgeAll()
  }
}
